Code/Framework/AzCore/AzCore/Memory/HphaSchema.cpp

```diff
@@ -2599,4 +2697,19 @@ namespace AZ
         return m_capacity;
     }
 
+    template<bool DebugAllocator>
+    size_t HphaSchemaBase<DebugAllocator>::GetMemoryGuardSize()
+    {
+        return HphaSchemaBase<DebugAllocator>::HpAllocator::MEMORY_GUARD_SIZE;
+    }
+
+    template<bool DebugAllocator>
+    size_t HphaSchemaBase<DebugAllocator>::GetFreeLinkSize()
+    {
+        return sizeof(typename HphaSchemaBase<DebugAllocator>::HpAllocator::free_link);
+    }
+
+    // explicitly instantiate both MallocSchemaBase classes
+    template class HphaSchemaBase<false>;
+    template class HphaSchemaBase<true>;
```

Comment:
The only thing that uses `HphachemaBase<true>` is the unit tests. With this change, we're adding 2 copies of the HPHA allocator code to the `AzCore` static library. What impact does that have on our build size? Does it get stripped at link time due to being unused?

Reply:
The size that it adds to the AzCore static library is is about 1 MiB in debug and 600 KiB in profile.

Here is the size of AzCore without the HphaSchema class template split
## Debug
* AzCore.lib - 705,068 KB
* AzCore.pdb - 44,484 KB

## Profile
* AzCore.lib - 267,993 KB
* AzCore.pdb - 46,180 KB

Here is the size of AzCore with the template split
## Debug
* AzCore.lib - 706,012 KB
* AzCore.pdb - 44,484 KB

## Profile
* AzCore.lib - 268,596 KB
* AzCore.pdb - 46,180 KB

At link time the symbol is stripped from all shared libraries and executables except for libAzCore.Tests.so
```bash
[user@arch o3de]$ nm -C build/linux_ninja/bin/profile/libAzCoreTestDLL.so | grep HphaSchemaBase 
0000000000023560 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::DeAllocate(void*, unsigned long, unsigned long)
0000000000023600 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::ReAllocate(void*, unsigned long, unsigned long)
0000000000023650 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::AllocationSize(void*)
00000000000236b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GarbageCollect()
00000000000237d0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetDebugConfig()
00000000000235b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Resize(void*, unsigned long)
0000000000023800 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Destroy()
00000000000234f0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Allocate(unsigned long, unsigned long, int, char const*, char const*, int, unsigned int)
0000000000023480 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
0000000000023420 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
00000000000236e0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::NumAllocatedBytes() const
0000000000023740 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxAllocationSize() const
00000000000237a0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetUnAllocatedMemory(bool) const
0000000000023770 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxContiguousAllocationSize() const
0000000000023710 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Capacity() const
00000000000d7698 d typeinfo for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000a910f r typeinfo name for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000d75b8 d vtable for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
[mcgarrah@arch o3de]$ nm -C build/linux_ninja/bin/profile/libAzCoreTestDLL.so | grep HphaSchema
0000000000023560 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::DeAllocate(void*, unsigned long, unsigned long)
0000000000023600 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::ReAllocate(void*, unsigned long, unsigned long)
0000000000023650 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::AllocationSize(void*)
00000000000236b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GarbageCollect()
00000000000237d0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetDebugConfig()
00000000000235b0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Resize(void*, unsigned long)
0000000000023800 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Destroy()
00000000000234f0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Allocate(unsigned long, unsigned long, int, char const*, char const*, int, unsigned int)
0000000000023480 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
0000000000023420 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::~SimpleSchemaAllocator()
00000000000236e0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::NumAllocatedBytes() const
0000000000023740 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxAllocationSize() const
00000000000237a0 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetUnAllocatedMemory(bool) const
0000000000023770 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::GetMaxContiguousAllocationSize() const
0000000000023710 t AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>::Capacity() const
00000000000d7698 d typeinfo for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000a910f r typeinfo name for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
00000000000d75b8 d vtable for AZ::SimpleSchemaAllocator<AZ::HphaSchema, AZ::HphaSchemaBase<false>::Descriptor, false, false>
```